server/test/repository/ProgramRepositoryTest.java

```diff
@@ -305,6 +311,153 @@ private Application makeApplicationWithName(
     return application;
   }
 
+  private static final StatusDefinitions.Status FIRST_STATUS =
+      StatusDefinitions.Status.builder()
+          .setStatusText("First status")
+          .setLocalizedStatusText(LocalizedStrings.withDefaultValue("First status"))
+          .build();
+
+  private static final StatusDefinitions.Status SECOND_STATUS =
+      StatusDefinitions.Status.builder()
+          .setStatusText("Second status")
+          .setLocalizedStatusText(LocalizedStrings.withDefaultValue("Second status"))
+          .build();
+
+  private static final StatusDefinitions.Status THIRD_STATUS =
+      StatusDefinitions.Status.builder()
+          .setStatusText("Third status")
+          .setLocalizedStatusText(LocalizedStrings.withDefaultValue("Third status"))
+          .build();
+
+  @Test
+  public void getApplicationsForAllProgramVersions_filterByStatus() throws Exception {
+    Program program =
+        ProgramBuilder.newActiveProgram("test program", "description")
+            .withStatusDefinitions(
+                new StatusDefinitions(ImmutableList.of(FIRST_STATUS, SECOND_STATUS, THIRD_STATUS)))
+            .build();
+
+    Account adminAccount = resourceCreator.insertAccountWithEmail("admin@example.com");
+
+    Application firstStatusApplication =
+        resourceCreator.insertActiveApplication(resourceCreator.insertApplicant(), program);
+    createStatusEvents(adminAccount, firstStatusApplication, FIRST_STATUS);
+
+    Application secondStatusApplication =
+        resourceCreator.insertActiveApplication(resourceCreator.insertApplicant(), program);
+    createStatusEvents(adminAccount, secondStatusApplication, SECOND_STATUS);
+
+    Application thirdStatusApplication =
+        resourceCreator.insertActiveApplication(resourceCreator.insertApplicant(), program);
+    // Create a few status events before-hand to ensure that the latest status is used.
+    createStatusEvents(
+        adminAccount, thirdStatusApplication, FIRST_STATUS, SECOND_STATUS, THIRD_STATUS);
+
+    Application noStatusApplication =
+        resourceCreator.insertActiveApplication(resourceCreator.insertApplicant(), program);
+
+    Application backToNoStatusApplication =
+        resourceCreator.insertActiveApplication(resourceCreator.insertApplicant(), program);
+    // Application has transitioned through statuses and arrived back at an empty status (indicated
+    // by null).
+    createStatusEvents(adminAccount, backToNoStatusApplication, SECOND_STATUS, null);
```

Comment:
1. We shouldn't use null unless forced too so use Optional not null, or another pattern.  We're following the google java guides which dissuade null unless there's no option.

2. you only use the status text off of the status, so just pass that and empty string as needed directly.

Reply:
Good call to not use null. Unless you feel strongly, I have a personal preference to pass an `Optional<StatusDefinitions.Status>` rather than a string since I think passing the type better documents the intended type of parameter than an arbitrary string (even though a string is sufficient).